File: elixir/0004.exs
defmodule Problem4 do
  def reverse_digits(n) when n == 0 do
    0
  end

  def reverse_digits(n) do
    rem(n, 10)*(:math.pow(10, :math.log10(n) |> trunc) |> trunc) + reverse_digits(div(n, 10))
  end

  def is_palindrome?(n) do
    n == reverse_digits n
  end

  @spec Enum &&& Enum :: Enum
  def x &&& y do
    # Cartesian product
    for i <- x, j <- y, do: [i, j]
  end

  def product it do
      Enum.reduce(it,  &*/2)
  end

  def largest_palindrome seq do
    seq |> Enum.filter(&is_palindrome?/1) |> Enum.max
  end

  def solution do
    (1..999 &&& 1..999) |> Enum.map(&product/1) |> largest_palindrome
  end
end


IO.puts "The solution is #{Problem4.solution}"

File: python/0004.py
from itertools import product

solution4 = max(
    i*j
    for i, j
    in product(range(100, 1000), range(100, 1000))
    if i*j == int(str(i*j)[::-1])
)

# update with := syntax in 3.8
# largest_palindrome = max(
#     (P := i*j)
#     for i, j
#     in product(range(100, 1000), range(100, 1000))
#     if P == int(str(P)[::-1])
# )

print(solution4)